test/addressBook.test.ts

```diff
@@ -0,0 +1,305 @@
+
+      oracle = await UpgradeableContractV1.at(await addressBook.getOracle())
+
+      assert.equal((await oracle.CONTRACT_REVISION()).toString(), '1', 'Oracle implementation version mismatch')
+    })
+  })
+
+  describe('Set weth', () => {
+    it('should revert adding weth address from non-owner address', async () => {
+      await expectRevert(addressBook.setWeth(weth.address, {from: random}), 'Ownable: caller is not the owner')
+    })
+
+    it('should set weth address', async () => {
+      await addressBook.setWeth(weth.address, {from: owner})
+
+      assert.equal(await addressBook.getWeth(), weth.address, 'WETH address mismatch')
+    })
+  })
+
+  describe('Set arbitrary address', () => {
+    const modulekey = web3.utils.fromAscii('newModule')
+    let module: UpgradeableContractV1Instance
+
+    before(async () => {
+      module = await UpgradeableContractV1.new()
+    })
+
+    it('should revert adding arbitrary key:address from non-owner address', async () => {
+      await expectRevert(
+        addressBook.updateImpl(modulekey, module.address, {from: random}),
+        'Ownable: caller is not the owner',
+      )
+    })
+
+    it('should set new module key and address', async () => {
+      await addressBook.updateImpl(modulekey, module.address, {from: owner})
+
+      const proxy: OwnedUpgradeabilityProxyInstance = await OwnedUpgradeabilityProxy.at(
+        await addressBook.getAddress(modulekey),
+      )
+      const implementationAddress = await proxy.implementation()
+
+      assert.equal(module.address, implementationAddress, 'New module implementation address mismatch')
+    })
+  })
+
+  describe('Upgrade contract', async () => {
+    let v2Controller: UpgradeableContractV2Instance
+
+    before(async () => {
+      v2Controller = await UpgradeableContractV2.new()
+    })
+
+    it('should upgrade contract from V1 to V2', async () => {
+      const v1Proxy: OwnedUpgradeabilityProxyInstance = await OwnedUpgradeabilityProxy.at(
+        await addressBook.getController(),
+      )
+      const v1ImplementationAddress = await v1Proxy.implementation()
+
+      assert.notEqual(v2Controller.address, v1ImplementationAddress, 'Controller V1 implementation address mismatch')
+
+      await addressBook.setController(v2Controller.address, {from: owner})
+
+      const v2Proxy: OwnedUpgradeabilityProxyInstance = await OwnedUpgradeabilityProxy.at(
+        await addressBook.getController(),
+      )
+      const v2ImplementationAddress = await v2Proxy.implementation()
+
+      assert.equal(v2Controller.address, v2ImplementationAddress, 'Controller V2 implementation address mismatch')
+      assert.equal(v1Proxy.address, v2Proxy.address, 'Controller proxy address mismatch')
+
+      v2Controller = await UpgradeableContractV2.at(await addressBook.getController())
+
+      assert.equal(
+        (await v2Controller.CONTRACT_REVISION()).toString(),
+        '2',
+        'Controller implementation version mismatch',
```

Comment:
So this is testing how you would upgrade the controller from v1 to v2? The address book itself is not upgradable right?

Reply:
yes this is testing upgrading from V1 to V2, the AddressBook is not upgradeable and does not need to be upgradeable in my opinion.